embassy-sync/src/mutex.rs

```diff
@@ -180,3 +198,122 @@ where
         unsafe { &mut *(self.mutex.inner.get()) }
     }
 }
+
+/// A handle to a held `Mutex` that has had a function applied to it via [`MutexGuard::map`] or
+/// [`MappedMutexGuard::map`].
+///
+/// This can be used to hold a subfield of the protected data.
+#[clippy::has_significant_drop]
+pub struct MappedMutexGuard<'a, M, T, U>
+where
+    M: RawMutex,
+    T: ?Sized,
+    U: ?Sized,
+{
+    mutex: &'a Mutex<M, T>,
```

Comment:
I think you can avoid one generic param by storing a reference to just the `state` field (of type `&'a BlockingMutex<M, RefCell<State>>`) instead of the whole `Mutex`, which would simplify things a bit.

Reply:
Done. Although I am not sure (again) about soundness

Reply:
If fact, I am not sure about anything 👍